### PR TITLE
Experimental: Env Setup Script for terminal

### PR DESCRIPTION
**Pull request by rohit-kumar-j**
Added "environmentSetupScript" option in terminal execution. 
Does not yet work in overseer or quickfix list, but option is provided (I'm not sure how to use overseer 😅)

```json
    {
        "name": "Windows Clang 16.0",
        "environmentSetupScript": "& \"C:\\Program Files\\Microsoft Visual Studio\\2022\\Community\\VC\\Auxiliary\\Build\\vcvars64.bat\"",
        "generator": "Ninja",
        "linker" : "C:/Program Files/LLVM/bin/lld-link.exe",
        "compilers": {
            "C": "C:/Program Files/LLVM/bin/clang.exe",
            "CXX": "C:/Program Files/LLVM/bin/clang.exe"
        }
    }
```

Ref: https://github.com/microsoft/vscode-cmake-tools/issues/809#issuecomment-662166588

---

#### EDIT: 

Additionally, minor semantic terminal mode fixes.

### CODE REVIEW

**Comment by rohit-kumar-j** [comment]
Even if we delete the buffers, as cmake tools maintains internal state, we can re-init the buffer with the env setup script

File: lua/cmake-tools/executors/terminal.lua [hunk 1]
```diff
@@ -449,7 +450,6 @@ function terminal.prepare_cmd_for_execute(executable, args, launch_path, wrap_ca
   -- executable = vim.fn.fnamemodify(executable, ":t")
 
   -- Launch form executable's build directory by default
-  launch_path = terminal.prepare_launch_path(launch_path)
```
Comment (rohit-kumar-j) [inline lua/cmake-tools/executors/terminal.lua:452]: Discussion : https://github.com/Civitasv/cmake-tools.nvim/issues/90#issuecomment-1703959109

**Comment by rohit-kumar-j** [comment]
Added minor fix for build directory name wrapping

File: lua/cmake-tools/executors/terminal.lua [hunk 2]
```diff
@@ -542,6 +542,20 @@ function terminal.run(cmd, env, args, cwd, opts)
   -- Reposition the terminal buffer, before sending commands
   local final_win_id = terminal.reposition(opts)
 
+  --- NOTE: env_script needs to be run only once if the terminal buffer does not already exist
+  --- We compare the old and the new id and only if they are not the same, plus if the terminal exists,
+  --    only then, we do not reinitialize the environment, else we reinit the env
+  if not terminal_already_exists or terminal.id_old ~= terminal.id then
```
Comment (Civitasv) [inline lua/cmake-tools/executors/terminal.lua:548]: This should ensure env_script is exists

**Comment by Civitasv** [comment]
Generally, the changes are okay.
Please make sure do not format the lines if you do not change the content.
And fix the format error. :smile:

File: lua/cmake-tools/executors/terminal.lua (hunk 2 above)
Comment (rohit-kumar-j) [reply lua/cmake-tools/executors/terminal.lua]: Done

**Comment by rohit-kumar-j** [comment]
> Generally, the changes are okay. Please make sure do not format the lines if you do not change the content. And fix the format error. 😄

Sorry about the formatting, I will change to use stylua instead of lua_ls. Do you have any pointers for setting this up?

**Comment by rohit-kumar-j** [comment]
Sorry for so many formatting errors😅

**Comment by Civitasv** [comment]
You should setup pre-commit, see https://pre-commit.com/.

**Comment by rohit-kumar-j** [comment]
> You should setup pre-commit, see https://pre-commit.com/.

This is good practice. I tried to set this up, but keep getting errors. Seems like there is an issue with running this on windows...

https://github.com/pre-commit/pre-commit/issues/2263

<details>
<Summary>Trace</Summary>

```bash
(venv) PS C:\Users\jsk71\AppData\Local\nvim-data\lazy\cmake-tools.nvim> pre-commit run --all-fils
usage: pre-commit [-h] [-V]
                  {autoupdate,clean,gc,init-templatedir,install,install-hooks,migrate-config,run,sample-config,try-repo,uninstall,validate-config,validate-manifest,help,hook-impl}
                  ...
pre-commit: error: unrecognized arguments: --all-fils
(venv) PS C:\Users\jsk71\AppData\Local\nvim-data\lazy\cmake-tools.nvim> pre-commit run --all-files
[INFO] Installing environment for https://github.com/JohnnyMorganz/StyLua.
[INFO] Once installed this environment will be reused.
[INFO] This may take a few minutes...
An unexpected error has occurred: CalledProcessError: command: ('C:\\Users\\jsk71\\.cache\\pre-commit\\repo8i96iou0\\py_env-python3.11\\Scripts\\python.EXE', '-mpip', 'install', '.')
return code: 1
stdout:
    Processing c:\users\jsk71\.cache\pre-commit\repo8i96iou0
      Installing build dependencies: started
      Installing build dependencies: finished with status 'done'
      Getting requirements to build wheel: started
      Getting requirements to build wheel: finished with status 'done'
      Preparing metadata (pyproject.toml): started
      Preparing metadata (pyproject.toml): finished with status 'done'
    Building wheels for collected packages: pseudo
      Building wheel for pseudo (pyproject.toml): started
      Building wheel for pseudo (pyproject.toml): finished with status 'error'
    Failed to build pseudo
stderr:
      error: subprocess-exited-with-error

      Building wheel for pseudo (pyproject.toml) did not run successfully.
      exit code: 1

      [20 lines of output]
      Starting new HTTPS connection (1): api.github.com:443
      https://api.github.com:443 "GET /repos/JohnnyMorganz/StyLua/releases HTTP/1.1" 200 None
      Traceback (most recent call last):
        File "C:\Users\jsk71\.cache\pre-commit\repo8i96iou0\py_env-python3.11\Lib\site-packages\pip\_vendor\pyproject_hooks\_in_process\_in_process.py", line 353, in <module>
          main()
        File "C:\Users\jsk71\.cache\pre-commit\repo8i96iou0\py_env-python3.11\Lib\site-packages\pip\_vendor\pyproject_hooks\_in_process\_in_process.py", line 335, in main
          json_out['return_val'] = hook(**hook_input['kwargs'])
                                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
        File "C:\Users\jsk71\.cache\pre-commit\repo8i96iou0\py_env-python3.11\Lib\site-packages\pip\_vendor\pyproject_hooks\_in_process\_in_process.py", line 251, in build_wheel
          return _build_backend().build_wheel(wheel_directory, config_settings,
                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
        File "C:\Users\jsk71\AppData\Local\Temp\pip-build-env-v5npgoqe\overlay\Lib\site-packages\pseudo_builder.py", line 133, in build_wheel
          files = download(metadata)
                  ^^^^^^^^^^^^^^^^^^
        File "C:\Users\jsk71\AppData\Local\Temp\pip-build-env-v5npgoqe\overlay\Lib\site-packages\pseudo_builder.py", line 25, in download
          asset = rg.match_asset(
                  ^^^^^^^^^^^^^^^
        File "C:\Users\jsk71\AppData\Local\Temp\pip-build-env-v5npgoqe\overlay\Lib\site-packages\release_gitter.py", line 268, in match_asset
          raise ValueError(
      ValueError: Could not find asset named {'stylua-windows-AMD64.zip'} on release v0.17.1
      [end of output]

      note: This error originates from a subprocess, and is likely not a problem with pip.
      ERROR: Failed building wheel for pseudo
    ERROR: Could not build wheels for pseudo, which is required to install pyproject.toml-based projects
Check the log at C:\Users\jsk71\.cache\pre-commit\pre-commit.log
(venv) PS C:\Users\jsk71\AppData\Local\nvim-data\lazy\cmake-tools.nvim>
```

</details>

**Comment by Civitasv** [comment]
I've fixed for you.

File: lua/cmake-tools/executors/terminal.lua [hunk 3]
```diff
@@ -542,6 +542,20 @@ function terminal.run(cmd, env, args, cwd, opts)
...
+    terminal.id_old = terminal.id
+    terminal.send_data_to_terminal(buffer_idx, env_script, {
+      win_id = final_win_id,
+      split_direction = opts.split_direction,
+      split_size = opts.split_size,
+      start_insert = opts.start_insert_in_other_tasks,
+      focus_on_main_terminal = opts.focus_on_main_terminal,
+    })
+  end
+
```
Comment (Civitasv) [inline lua/cmake-tools/executors/terminal.lua:558]: Why should we run the script if the terminal already exists? I think we should only run the script when the terminal did not exist or terminal.id_old not equals terminal.id.

File: lua/cmake-tools/init.lua [hunk 4]
```diff
@@ -197,6 +209,15 @@ function cmake.generate(opt, callback)
...
 
+  -- if exists cmake-kits.json, kit is used to set
+  -- environmental variables and args.
+  local kits_config = kits.parse(const.cmake_kits_path, config.cwd)
+  if kits_config and not config.kit then
+    return cmake.select_kit(function()
+      cmake.generate(opt, callback)
+    end)
+  end
+
```
Comment (Civitasv) [inline lua/cmake-tools/init.lua:220]: Why should move this part here?

File: lua/cmake-tools/init.lua [hunk 5]
```diff
@@ -265,7 +307,7 @@ function cmake.clean(callback)
     return log.error(result.message)
   end
 
-  local args = { "--build", config.build_directory.filename, "--target", "clean" }
+  local args = { "--build", '"' .. config.build_directory.filename .. '"', "--target", "clean" }
```
Comment (Civitasv) [inline lua/cmake-tools/init.lua:310]: The "" is for Windows?

Comment (rohit-kumar-j) [reply lua/cmake-tools/init.lua:310]: Yes

File: lua/cmake-tools/executors/terminal.lua (hunk 3 above)
Comment (rohit-kumar-j) [reply lua/cmake-tools/executors/terminal.lua:558]: Ah... I had an error for this, when I used `:bclose` or `:bdelete` on the terminal buffer and vim threw an error for this, saying buffer not found. That is why I made it something like this...

> This should ensure env_script is exists
>>  if not terminal_already_exists or terminal.id_old ~= terminal.id then

However, I was not able to reproduce the error, so I noted it down and added extra checking: `if not terminal_already_exists or ... `

File: lua/cmake-tools/init.lua (hunk 4 above)
Comment (rohit-kumar-j) [reply lua/cmake-tools/init.lua:220]: In the `select_build_type()` funtion -> `vim.ui.select()`[line 905], the `cmake_generate()` or the `callback()` branches do not pick-up the changes from the CMakeKits.json and do not detect the modified envScript, if you have modified the `.json` file and then re-run /re-opened vim.

Moving this before the generate callback makes sure that this does not happen, or atleast that is my idea. This is not thoroughly tested, as I was only working with `:CMakeGenerate` each time to test this feature, and not with other startup commands, like `:CMakeRun` or `:CMakeBuild`.

PS: Although there is a cmake.clean() command before this(line 115), I don't think running the envScript for clean is necessary, otherwise, it would be best to move it there.

File: lua/cmake-tools/init.lua (hunk 5 above)
Comment (rohit-kumar-j) [reply lua/cmake-tools/init.lua:310]: This isn't really necessary, It messes up the file paths. We can just ask users to simply not use paths with space which get translated to: `'C:\Program Files\....'` Otherwise we need to do spearate file path handling operations for windows, and `:CMakeArgs' (https://github.com/Civitasv/cmake-tools.nvim/issues/90#issuecomment-1703959109) would not work on Windows as of now.

File: lua/cmake-tools/executors/terminal.lua (hunk 3 above)
Comment (rohit-kumar-j) [reply lua/cmake-tools/executors/terminal.lua]: Update: 

Nope, if the  "not" is not present, the env script runs each time, so we need:

```bash
if not terminal_already_exists or terminal.id_old ~= terminal.id then
  ...
end
```

**Comment by rohit-kumar-j** [comment]
Two fixes, after testing

File: lua/cmake-tools/init.lua (hunk 4 above)
Comment (Civitasv) [reply lua/cmake-tools/init.lua:220]: I don't quite understand. Why does kits have relationship with build type? Can you give an example?

**Comment by Civitasv** [comment]
And you should also add `env_script` arg to the arguments of overseer_executor.run to make sure it follows `executor.run`.

File: lua/cmake-tools/init.lua (hunk 4 above)
Comment (rohit-kumar-j) [reply lua/cmake-tools/init.lua]: Sorry for the change, it has nothing to with build_type, ... was a mistake on my end... reverting change

## TL;DR
When I was developing the envScript thing, on my working branch I had not yet created a save method in sesions to save the envScript. So, when doing `:CmakeGenerate` on a directory which did not (yet) have a sessions file in the cache, the envScript was loaded and ran correctly. 

However, upon modifying the `CMakeKits.json` to remove(or modify) the envSvript, and reopening nvim the same directory, the changes in the file were not being detected from the modified `CMakeKits.json`, so in frustration, I moved this above the `select_build_type()` function, which solved this problem (don't exactly know how... probably had something to do with the saving of the sessions file).

Now after reverting the change and re-testing this, the bug is ... gone :D

**Comment by rohit-kumar-j** [comment]
On this note:
> Sorry for the change, it has nothing to with build_type, ... was a mistake on my end... reverting change
> 
> ## TL;DR
> When I was developing the envScript thing, on my working branch I had not yet created a save method in sesions to save the envScript. So, when doing `:CmakeGenerate` on a directory which did not (yet) have a sessions file in the cache, the envScript was loaded and ran correctly.
> 
> However, upon modifying the `CMakeKits.json` to remove(or modify) the envSvript, and reopening nvim the same directory, the changes in the file were not being detected from the modified `CMakeKits.json`, so in frustration, I moved this above the `select_build_type()` function, which solved this problem (don't exactly know how... probably had something to do with the saving of the sessions file).
> 
> Now after reverting the change and re-testing this, the bug is ... gone :D

Should we support live reloading of the `config` table upon `BufWritePre` for kits, and presets... Similar to #58  's AutoRun of `:CMakeGenerate` upon `BufWritePre`? (In another PR)

Something like: `cmake.create_rescan_kits_and_presets_changes_on_save_autocmd()`, which can be enabled by default, but does not run any commands to the terminal. Similar to: `cmake.create_regenerate_on_save_autocmd()`?

**Comment by rohit-kumar-j** [comment]
Rebased

File: lua/cmake-tools/init.lua (hunk 4 above)
Comment (Civitasv) [reply lua/cmake-tools/init.lua]: Interesting, lol.

**Comment by Civitasv** [comment]
> Something like: `cmake.create_rescan_kits_and_presets_changes_on_save_autocmd()`, which can be enabled by default, but does not run any commands to the terminal. Similar to: `cmake.create_regenerate_on_save_autocmd()`?

I think it has been supported, See:
https://github.com/Civitasv/cmake-tools.nvim/blob/9ffcf4f4a31215b24fc776f0b85563ebdad5cf89/lua/cmake-tools/init.lua#L1377-L1384

File: lua/cmake-tools/executors/terminal.lua [hunk 6]
```diff
@@ -170,7 +171,7 @@ function terminal.send_data_to_terminal(buffer_idx, cmd, opts)
     -- do nothing
   end
 
-  if opts and not (opts.focus_on_launch_terminal or opts.focus_on_main_terminal) then
+  if opts and (opts.focus_on_launch_terminal or opts.focus_on_main_terminal) then
```
Comment (Civitasv) [inline lua/cmake-tools/executors/terminal.lua:174]: Why did you remove the `not`?

**Comment by Civitasv** [comment]
Oher than comments above, LGTM.

File: lua/cmake-tools/executors/terminal.lua (hunk 6 above)
Comment (rohit-kumar-j) [reply lua/cmake-tools/executors/terminal.lua:176]: Slight bug here:

Sometimes there is a problem with buffers that are already open, which we do not want to focus on... like lets say we switch to nerdtree, and switch back to our `.cpp` file, and run `:CMake<...>` then it focuses on the nerdtree if that buffer is open... which can be annoying. So, a more robust system of detecting/switching buffers is needed like the buffer name/number and detection prior to `wincmd 'p'`.

**Comment by rohit-kumar-j** [comment]
> Why did you remove the `not`? To my understanding, if `opts. focus_on_launch_terminal or opts.focus_on_main_terminal` is not true, we should set current window to previous window, isn't it?

Sent fix 😃

File: lua/cmake-tools/executors/terminal.lua (hunk 6 above)
Comment (rohit-kumar-j) [reply lua/cmake-tools/executors/terminal.lua:176]: I modified this logic completely now 👀

File: lua/cmake-tools/executors/terminal.lua [hunk 7]
```diff
@@ -164,31 +165,49 @@ function terminal.send_data_to_terminal(buffer_idx, cmd, opts)
...
   else
-    vim.api.nvim_set_current_win(opts.win_id)
+    -- We want to focus on our currently focused window and not ther cmake terminal
+    local name = vim.api.nvim_buf_get_name(vim.api.nvim_win_get_buf(0))
+    local basename = vim.fn.fnamemodify(name, ":t")
+    if opts and (basename:sub(1, #opts.prefix) == opts.prefix) then -- If currently focused buffer is cmake buffer then ...
+      -- Now we check again if the buffer needs to be focused as the user might be scrolling
+      -- a cmake buffer and execute a :CMakeCommand, so we do not want to move their
+      -- cursor out of the cmake buffer, as it can be annoying
+      if opts and (opts.focus_on_launch_terminal or opts.focus_on_main_terminal) then
```
Comment (Civitasv) [inline lua/cmake-tools/executors/terminal.lua:191]: Do you forger a `not`?
To my understanding, it should be `if opts and not (opts.focus_on_launch_terminal or opts.focus_on_main_terminal)`

Comment (rohit-kumar-j) [reply lua/cmake-tools/executors/terminal.lua:191]: Yes, this is better. Updated.

**Comment by rohit-kumar-j** [comment]
Rebased

**Comment by Civitasv** [comment]
LGTM.
